schema-engine/commands/src/commands/diff.rs

```diff
@@ -0,0 +1,192 @@
+use crate::{
+    CoreError, SchemaContainerExt,
+    core_error::CoreResult,
+    dialect_for_provider, extract_namespaces,
+    json_rpc::types::{DiffParams, DiffResult, DiffTarget, UrlContainer},
+};
+use enumflags2::BitFlags;
+use json_rpc::types::MigrationList;
+use psl::SourceFile;
+use schema_connector::{
+    ConnectorError, DatabaseSchema, ExternalShadowDatabase, Namespaces, SchemaConnector, SchemaDialect,
+};
+
+pub async fn diff(params: DiffParams, connector: &mut dyn SchemaConnector) -> CoreResult<DiffResult> {
+    // In order to properly handle MultiSchema, we need to make sure the preview feature is
+    // correctly set, and we need to grab the namespaces from the Schema, if any.
+    // Note that currently, we union all namespaces and preview features. This may not be correct.
+    // TODO: This effectively reads and parses (parts of) the schema twice: once here, and once
+    // below, when defining 'from'/'to'. We should revisit this.
+    let (namespaces, preview_features) =
+        namespaces_and_preview_features_from_diff_targets(&[&params.from, &params.to])?;
+
+    let from = diff_target_to_dialect(
+        &params.from,
+        params.shadow_database_url.as_deref(),
+        connector,
+        namespaces.clone(),
+        preview_features,
+    )
+    .await?;
+    let to = diff_target_to_dialect(
+        &params.to,
+        params.shadow_database_url.as_deref(),
+        connector,
+        namespaces,
+        preview_features,
+    )
+    .await?;
+
+    // The `from` connector takes precedence, because if we think of diffs as migrations, `from` is
+    // the target where the migration would be applied.
+    //
+    // TODO: make sure the connectors are the same in from and to.
```

Comment:
```suggestion
    // TODO: make sure the connectors are the same in from and to.
```
I wonder if this is actually necessary, might be a 'feature'

Reply:
That feature doesn't work right now, but sure, it could evolve. This is copied verbatim from the original `diff` implementation, btw